rest/import_test.go

```diff
@@ -698,6 +698,196 @@ func TestMigrateWithExternalRevisions(t *testing.T) {
 	assert.Equals(t, len(doc.Meta.RevTree.BodyMap), 2)
 }
 
+// Write a doc via SDK with an expiry value.  Verify that expiry is preserved when doc is imported via DCP feed
+func TestXattrFeedBasedImportPreservesExpiry(t *testing.T) {
+
+	SkipImportTestsIfNotEnabled(t)
+
+	rt := RestTester{
+		SyncFn: `function(doc, oldDoc) { channel(doc.channels) }`,
+		DatabaseConfig: &DbConfig{
+			ImportDocs: "continuous",
+		},
+	}
+	defer rt.Close()
+	bucket := rt.Bucket()
+
+	rt.SendAdminRequest("PUT", "/_logging", `{"Import+":true, "CRUD+":true}`)
+
+	// 1. Create docs via the SDK with expiry set
+	mobileKey := "TestXattrImportPreservesExpiry"
+	mobileBody := make(map[string]interface{})
+	mobileBody["type"] = "mobile"
+	mobileBody["channels"] = "ABC"
+
+	// Write directly to bucket with an expiry
+	expiryUnixEpoch := time.Now().Add(time.Second * 30).Unix()
+	_, err := bucket.Add(mobileKey, uint32(expiryUnixEpoch), mobileBody)
+	assertNoError(t, err, "Error writing SDK doc")
+
+	// Wait until the change appears on the changes feed to ensure that it's been imported by this point
+	changes, err := rt.WaitForChanges(1, "/db/_changes", "", true)
+	assertNoError(t, err, "Error waiting for changes")
+
+	log.Printf("changes: %+v", changes)
+	changeEntry := changes.Results[0]
+	assert.Equals(t, changeEntry.ID, mobileKey)
+
+	// Double-check to make sure that it's been imported by checking the Sync Metadata in the xattr
+	assertXattrSyncMetaRevGeneration(t, bucket, mobileKey, 1)
+
+	// Verify the expiry has been preserved after the import
+	gocbBucket := bucket.(*base.CouchbaseBucketGoCB)
+	expiry, err := gocbBucket.GetExpiry(mobileKey)
+	assertNoError(t, err, "Error calling GetExpiry()")
+	assert.True(t, expiry == uint32(expiryUnixEpoch))
+
+}
+
+// Test migration of a 1.5 doc that has an expiry value.
+func TestFeedBasedMigrateWithExpiry(t *testing.T) {
+
+	SkipImportTestsIfNotEnabled(t)
+
+	rt := RestTester{
+		SyncFn: `function(doc, oldDoc) { channel(doc.channels) }`,
+		DatabaseConfig: &DbConfig{
+			ImportDocs: "continuous",
+		},
+	}
+	defer rt.Close()
+	bucket := rt.Bucket()
+
+	rt.SendAdminRequest("PUT", "/_logging", `{"Import+":true, "CRUD+":true}`)
+
+	// Write doc in SG format directly to the bucket
+	key := "TestFeedBasedMigrateWithExpiry"
+
+	// Create via the SDK with sync metadata intact
+	expirySeconds := time.Second * 30
+	syncMetaExpiry := time.Now().Add(expirySeconds)
+	bodyString := expiringRawDocWithSyncMeta(syncMetaExpiry)
+	_, err := bucket.Add(key, 0, []byte(bodyString))
```

Comment:
Does it make sense to add the doc with an expiry, in addition to having an expiry in the `_sync` metadata, and ensure that it chooses the correct one?

Reply:
The expiry in the _sync metadata shouldn't be used as the source of truth - it's really just a lightweight way to retrieve expiry in case it's queried via the SG REST API.  In fact we could review whether it could be removed altogether now that we have GetMeta support.

The CBS expiry on the doc should always be used when working w/ expiry, even for migration.

Reply:
Fixed